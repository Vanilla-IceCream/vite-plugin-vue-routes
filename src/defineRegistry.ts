import path from 'path';
import { glob } from 'glob';
import { MagicString, getTransformResult, isCallOf, parseSFC, walkAST } from '@vue-macros/common';

import { hasTs } from './utils';

export default async (code: string, id: string) => {
  if (path.basename(id) !== 'Registry.vue') return;

  const hasMacro = code.includes('defineRegistry');

  const layouts = await glob(`src/layouts/*.vue`);
  const hasLayout = Boolean(layouts.length);

  const middlewareFiles = await glob(`src/middleware/*.{ts,js}`);
  const hasMiddleware = Boolean(middlewareFiles.length);

  if (!hasMacro && !hasLayout) return;

  const s = new MagicString(code);

  if (!hasMacro && hasLayout) {
    // injectLayout({ code, layout: 'Default', s });

    if (hasMiddleware) {
      if (middlewareFiles.some((file) => path.basename(file).includes('default'))) {
        injectMiddleware({ code, middleware: ['default'], s });
      }
    }
  }

  if (hasMacro) {
    const { scriptSetup, getSetupAst } = parseSFC(code, id);
    if (!scriptSetup) return;

    const setupAst = getSetupAst()!;
    const offset = scriptSetup.loc.start.offset;

    walkAST(setupAst, {
      enter(node) {
        if (isCallOf(node, 'defineRegistry')) {
          const match = code.match(/defineRegistry\(([\s\S]*?)\)/);

          const defineRegistryStr = match?.[1].trim();
          const defineRegistryObj = Function(`'use strict'; return (${defineRegistryStr})`)();
          const { layout, middleware } = JSON.parse(JSON.stringify(defineRegistryObj));

          const hasLayoutVal = typeof layout === 'string' && !!layout;
          const hasMiddlewareVal = !!middleware?.length;

          if (hasLayoutVal && !hasMiddlewareVal) {
            s.overwriteNode(node, '', { offset });

            // injectLayout({ code, layout, s });

            if (hasMiddleware) {
              if (
                !!middlewareFiles.filter((file) =>
                  layout.toLowerCase().includes(path.parse(file).name.toLowerCase()),
                ).length
              ) {
                injectMiddleware({ code, middleware: [layout.toLowerCase()], s });
              }
            }
          }

          if (!hasLayoutVal && hasMiddlewareVal) {
            s.overwriteNode(node, '', { offset });

            // injectLayout({ code, layout: 'Default', s });
            injectMiddleware({ code, middleware: ['default', ...middleware], s });
          }

          if (hasLayoutVal && hasMiddlewareVal) {
            s.overwriteNode(node, '', { offset });

            // injectLayout({ code, layout, s });

            if (
              !!middlewareFiles.filter((file) =>
                layout.toLowerCase().includes(path.parse(file).name.toLowerCase()),
              ).length
            ) {
              injectMiddleware({ code, middleware: [layout.toLowerCase(), ...middleware], s });
            } else {
              injectMiddleware({ code, middleware, s });
            }
          }

          if (!hasLayoutVal && !hasMiddlewareVal) {
            s.overwriteNode(node, '', { offset });
          }
        }
      },
    });
  }

  return getTransformResult(s, id);
};

interface Injector {
  code: string;
  layout: string;
  middleware: string[];
  s: MagicString;
}

// function injectLayout({ code, layout, s }: Omit<Injector, 'middleware'>) {
//   const _layout = layout.split('@');

//   if (code.includes('</script>')) {
//     s.replace(
//       /<\/script>/i,
//       _layout
//         .map(
//           (layoutName) =>
//             `import ${capitalize(layoutName)}Layout from '~/layouts/${capitalize(
//               layoutName,
//             )}.vue';`,
//         )
//         .join('') + '</script>',
//     );
//   } else {
//     s.prepend(
//       '<script setup>' +
//         _layout
//           .map(
//             (layoutName) =>
//               `import ${capitalize(layoutName)}Layout from '~/layouts/${capitalize(
//                 layoutName,
//               )}.vue';`,
//           )
//           .join('') +
//         '</script>',
//     );
//   }

//   s.replace(
//     /<template>([\s\S]+)<\/template>/,
//     '<template>' +
//       _layout.map((layoutName) => `<${capitalize(layoutName)}Layout>`).join('') +
//       '$1' +
//       _layout
//         .reverse()
//         .map((layoutName) => `</${capitalize(layoutName)}Layout>`)
//         .join('') +
//       '</template>',
//   );
// }

function injectMiddleware({ code, middleware, s }: Omit<Injector, 'layout'>) {
  const scriptLang = hasTs(code) ? `<script lang="ts">` : '<script>';

  let importMiddleware = '';

  for (let i = 0; i < middleware.length; i++) {
    const mod = middleware[i];
    importMiddleware += `import ${mod}Middleware from '~/middleware/${mod}';`;
  }

  s.prepend(`
    ${scriptLang}
    ${importMiddleware}

    export default {
      async beforeRouteEnter(to, from) {
        for (const func of [${middleware.map((m) => `${m}Middleware`)}]) {
          const result = await func(to, from);
          if (result !== true) return result;
        }

        return true;
      },
    };
    </script>
  `);
}
