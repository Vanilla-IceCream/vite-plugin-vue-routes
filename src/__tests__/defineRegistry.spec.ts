import { test, expect, vi } from 'vitest';

import defineRegistry from '../defineRegistry';

vi.mock('glob');

test('defineRegistry', async () => {
  const glob = await import('glob');

  glob.glob = vi.fn((pattern) => {
    if (pattern === 'src/layouts/*.vue') {
      return ['src/layouts/Default.vue'];
    }

    if (pattern === 'src/middleware/*.{ts,js}') {
      return ['src/middleware/default.ts'];
    }

    return [];
  }) as any;

  const code = `
    <template>
      <div>Default</div>
    </template>
  `;

  const generated = await defineRegistry(code, 'Registry.vue');
  expect(generated).toMatchSnapshot();
});

test('defineRegistry - layout', async () => {
  const glob = await import('glob');

  glob.glob = vi.fn((pattern) => {
    if (pattern === 'src/layouts/*.vue') {
      return ['src/layouts/Violet.vue'];
    }

    if (pattern === 'src/middleware/*.{ts,js}') {
      return ['src/middleware/default.ts'];
    }

    return [];
  }) as any;

  const code = `
    <script lang="ts" setup>
    defineRegistry({
      layout: 'Violet',
    });
    </script>

    <template>
      <div>Violet</div>
    </template>
  `;

  const generated = await defineRegistry(code, 'Registry.vue');
  expect(generated).toMatchSnapshot();
});

test('defineRegistry - middleware', async () => {
  const glob = await import('glob');

  glob.glob = vi.fn((pattern) => {
    if (pattern === 'src/layouts/*.vue') {
      return ['src/layouts/Violet.vue'];
    }

    if (pattern === 'src/middleware/*.{ts,js}') {
      return ['src/middleware/violet.ts'];
    }

    return [];
  }) as any;

  const code = `
    <script lang="ts" setup>
    defineRegistry({
      middleware: ['violet'],
    });
    </script>

    <template>
      <div>Violet</div>
    </template>
  `;

  const generated = await defineRegistry(code, 'Registry.vue');
  expect(generated).toMatchSnapshot();
});
