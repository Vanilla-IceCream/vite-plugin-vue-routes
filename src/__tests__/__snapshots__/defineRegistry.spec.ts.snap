// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`defineRegistry - layout 1`] = `
{
  "code": "
    <script lang=\\"ts\\" setup>
    ;
    </script>

    <template>
      <div>Violet</div>
    </template>
  ",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AAC5B,CAAC,CAAC,CAAC,CAEG,CAAC;AACP,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACb;AACA,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACd,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACvB,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACf,CAAC",
    "names": [],
    "sources": [
      "Registry.vue",
    ],
    "sourcesContent": [
      "
    <script lang=\\"ts\\" setup>
    defineRegistry({
      layout: 'Violet',
    });
    </script>

    <template>
      <div>Violet</div>
    </template>
  ",
    ],
    "version": 3,
  },
}
`;

exports[`defineRegistry - middleware 1`] = `
{
  "code": "
    <script lang=\\"ts\\">
    import defaultMiddleware from '~/middleware/default';import violetMiddleware from '~/middleware/violet';

    export default {
      async beforeRouteEnter(to, from) {
        for (const func of [defaultMiddleware,violetMiddleware]) {
          const result = await func(to, from);
          if (result !== true) return result;
        }

        return true;
      },
    };
    </script>
  
    <script lang=\\"ts\\" setup>
    ;
    </script>

    <template>
      <div>Violet</div>
    </template>
  ",
  "map": SourceMap {
    "file": undefined,
    "mappings": ";;;;;;;;;;;;;;;EAAA;AACA,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AAC5B,CAAC,CAAC,CAAC,CAEG,CAAC;AACP,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACb;AACA,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACd,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACvB,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACf,CAAC",
    "names": [],
    "sources": [
      "Registry.vue",
    ],
    "sourcesContent": [
      "
    <script lang=\\"ts\\" setup>
    defineRegistry({
      middleware: ['violet'],
    });
    </script>

    <template>
      <div>Violet</div>
    </template>
  ",
    ],
    "version": 3,
  },
}
`;

exports[`defineRegistry 1`] = `
{
  "code": "
    <script>
    import defaultMiddleware from '~/middleware/default';

    export default {
      async beforeRouteEnter(to, from) {
        for (const func of [defaultMiddleware]) {
          const result = await func(to, from);
          if (result !== true) return result;
        }

        return true;
      },
    };
    </script>
  
    <template>
      <div>Default</div>
    </template>
  ",
  "map": SourceMap {
    "file": undefined,
    "mappings": ";;;;;;;;;;;;;;;EAAA;AACA,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACd,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACxB,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;AACf,CAAC",
    "names": [],
    "sources": [
      "Registry.vue",
    ],
    "sourcesContent": [
      "
    <template>
      <div>Default</div>
    </template>
  ",
    ],
    "version": 3,
  },
}
`;
